src/hardware/bk-1856d/protocol.c

```diff
@@ -0,0 +1,325 @@
+	if(!devc || !serial)
+		return;
+
+	devc->buffer_level = 0;
+	devc->samples_read = 0;
+	serial_flush(serial);
+
+	bk_1856d_select_input(sdi);
+
+	bk_1856d_set_gate_time(sdi);
+	bk_1856d_request_data(sdi);
+}
+
+static int bk_1856d_check_for_zero_message(struct dev_context *devc)
+{
+	int zero_found = 0;
+	int has_data = 0;
+	for(int idx = 0; idx < BK1856D_MSG_SIZE-1 ; ++idx)
+	{
+		if(devc->buffer[idx] == ' ')
+			continue;
+		else if(devc->buffer[idx] == '0' && zero_found == 0)
+		{
+			zero_found = 1;
+			continue;
+		}
+		else
+		{
+			has_data = 1;
+			break;
+		}
+	}
+	return !has_data;
+}
+
+static void bk_1856d_send_packet(const struct sr_dev_inst *sdi, double freq_value, int digits)
+{
+	struct dev_context *devc = sdi->priv;
+
+	struct sr_datafeed_packet packet;
+	struct sr_datafeed_analog analog;
+	struct sr_analog_encoding encoding;
+	struct sr_analog_meaning meaning;
+	struct sr_analog_spec spec;
+
+	if (!devc)
+		return;
+
+	sr_analog_init(&analog, &encoding, &meaning, &spec, digits);
+	analog.meaning->mq = SR_MQ_FREQUENCY;
+	analog.meaning->unit = SR_UNIT_HERTZ;
+	analog.meaning->channels = sdi->channels;
+	analog.num_samples = 1;
+	analog.data = &freq_value;
+	analog.encoding->unitsize = sizeof(freq_value);
+	analog.encoding->is_float = TRUE;
+
+	packet.type = SR_DF_ANALOG;
+	packet.payload = &analog;
+	sr_session_send(sdi, &packet);
+}
+
+static void bk_1856d_parse_message(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc = sdi->priv;
+	struct sr_serial_dev_inst *serial = sdi->conn;
+
+	double freq_value;
+	int digits;
+	char *endPtr, *dotPtr;
+
+	if (!devc || !serial)
+		return;
+
+	/* check for cr at end of message */
+	if(devc->buffer[BK1856D_MSG_SIZE-1] != '\xD')
+	{
+		sr_err("expected cr at end of message.");
+		devc->buffer_level = 0;
+		serial_flush(serial);
+		bk_1856d_select_input(sdi);
+		bk_1856d_set_gate_time(sdi);
+		bk_1856d_request_data(sdi);
+		return;
+	}
+
+	devc->buffer[BK1856D_MSG_SIZE-1] = 0; /* set trailing zero */
+
+	if(bk_1856d_check_for_zero_message(devc))
+	{
+		sr_spew("received 0 package");
+		devc->buffer_level = 0;
+		bk_1856d_request_data(sdi);
+		return;
+	}
+
+	sr_dbg("received msg: '%s'", devc->buffer);
+	freq_value = strtod(devc->buffer, &endPtr);
+	sr_dbg("parsed value: %f", freq_value);
+
+	if(strcmp(devc->buffer + BK1856D_MSG_NUMBER_SIZE+1, "Hz ") != 0)
+	{
+		sr_err("not a frequency returned");
+		devc->buffer_level = 0;
+		bk_1856d_select_input(sdi);
+		bk_1856d_set_gate_time(sdi);
+		bk_1856d_request_data(sdi);
+		return;
+	}
+
+	dotPtr = strchr(devc->buffer, '.');
+	if(dotPtr)
+		digits = endPtr - (dotPtr+1);
+	else
+		digits = endPtr - devc->buffer;
+
+	if(devc->buffer[BK1856D_MSG_NUMBER_SIZE] == 'M')
+		freq_value *= 1e6;
+	if(devc->buffer[BK1856D_MSG_NUMBER_SIZE] == 'k')
+		freq_value *= 1e3;
+
+	bk_1856d_send_packet(sdi, freq_value, digits);
+
+	if(!(devc->continuous))
+		devc->samples_read += 1;
+
+	if(devc->continuous || devc->samples_read < devc->limit_samples)
```

Comment:
`src/sw_limits.c` provides an API for this purpose and would also give you limiting by frames or by time for free.

Reply:
I just pushed the changes using sw_limit to my fork.
I think it doesn't make sense to limit by time because you have to set the gate-time for the measurement anyway.